Add `Store.load_from_stream` and `Store.load_from_data` methods

edmondchuc opened the pull request:
Fixes https://github.com/oxigraph/oxigraph/issues/218

This PR implements two new methods to the `Store` class as suggested in: https://github.com/oxigraph/oxigraph/issues/218#issuecomment-1193384035 along with tests.

edmondchuc commented on python/src/store.rs:
```diff
@@ -322,6 +322,163 @@ impl PyStore {
+    /// :type to_graph: NamedNode or BlankNode or DefaultGraph or None, optional
+    /// :raises ValueError: if the MIME type is not supported or the `to_graph` parameter is given with a quad file.
+    /// :raises SyntaxError: if the provided data is invalid.
+    /// :raises IOError: if an I/O error happens during a quad insertion.
```
@Tpt would this still raise an `IOError` anywhere?

Tpt replied:
No, it won't.

edmondchuc commented:
@Tpt in https://github.com/oxigraph/oxigraph/issues/218#issuecomment-1193384035 you mentioned deprecating `Store.load()`. What should this entail? Log some warnings of the deprecation?

edmondchuc commented:
**Action:** 
- [ ] Instead of `load_from_stream`, create a method named `load_from_file` which accepts a file path. 
- [ ] Change `load_from_data` to accept a python I/O object or a `str`/`bytes` buffer.

Tpt commented:
An other naming idea: the shorter `load_file` and `load_data`

Tpt commented:
An idea to avoid code duplication: create a shared method not exposed with pyo3 that takes for input type `impl BufRead + Send` and use it in the three implemented methods. But it's likely to be a bit tricky so feel free to ignore this idea and let me attempt it in a follow-up change.

edmondchuc commented:
Hi @Tpt I've added my implementation of `load_file` and `load_data`. The former is the same code as the current `load` method and the latter accepts input as either a `str` or `bytes`. 

I am not sure if what I have implemented is idiomatic rust/pyo3. Please let me know if it should be done differently.

edmondchuc commented:
> An idea to avoid code duplication: create a shared method not exposed with pyo3 that takes for input type impl BufRead + Send and use it in the three implemented methods.

I am keen to attempt this if what I've done so far makes sense.

Tpt commented:
> @Tpt in https://github.com/oxigraph/oxigraph/issues/218#issuecomment-1193384035 you mentioned deprecating Store.load(). What should this entail? Log some warnings of the deprecation?

I believe we want to use Python deprecation warning system. The PyO3 library seems to expose it using [`PyErr::warn`](https://docs.rs/pyo3/latest/pyo3/prelude/struct.PyErr.html#method.warn) but the documentation is a bit lacking in this area. It's something we maybe should investigate in a follow up PR.

> Hi @Tpt I've added my implementation of load_file and load_data. The former is the same code as the current load method and the latter accepts input as either a str or bytes.
> I am not sure if what I have implemented is idiomatic rust/pyo3. Please let me know if it should be done differently.

Thank you so much! The code looks good. I would maybe move the Python I/O objects support to the `load_data` method because it's more "data" than a "file".

> I am keen to attempt this if what I've done so far makes sense.

That would be amazing. It might also help with the moving of I/O objects to `load_data` easier by allowing to write code like `if is_string { self.load_inner(the str) } else { self.load_inner(the_io_object) }` so I would suggest to do this refactoring first. But if you find it too hard, I can easily take some time to make an attempt. Rust generics are not the easiest thing (it does not have the C++ templates footguns but making the compiler happy with what ones write often takes some efforts).

edmondchuc commented:
> Thank you so much! The code looks good. I would maybe move the Python I/O objects support to the load_data method because it's more "data" than a "file".

Awesome, I've added support for Python I/O objects to `Store.load_data`. This includes `RawIOBase`, `BufferedIOBase` and `TextIOBase`. `TextIOBase` is a new addition. I've also added tests for these I/O object types.

> That would be amazing. It might also help with the moving of I/O objects to load_data easier by allowing to write code like if is_string { self.load_inner(the str) } else { self.load_inner(the_io_object) } so I would suggest to do this refactoring first. But if you find it too hard, I can easily take some time to make an attempt. Rust generics are not the easiest thing (it does not have the C++ templates footguns but making the compiler happy with what ones write often takes some efforts).

I fumbled around with the code, still a bit uncertain what I'm doing in rust, but I had a go. It deviates a bit from what you suggested. Let me know if there are any issues with it.

> An idea to avoid code duplication: create a shared method not exposed with pyo3 that takes for input type impl BufRead + Send and use it in the three implemented methods.

I wasn't able to use `impl BufRead + Send`. Apparently python functions can't have generics, unless I misunderstood something. 

The result was a unified method called `Store.load_inner` which accepts `input` as `&[u8]`.

Tpt commented:
>>I wasn't able to use `impl BufRead + Send`. Apparently python functions can't have generics, unless I misunderstood something.
>
> The result was a unified method called `Store.load_inner` which accepts input as &[u8].

Thank you so much for giving it a try. We are moving forward! Indeed, python functions can't have generics, my suggestion was to build a function not exposed to python (i.e. not in the `#[pymethods]` impl block). I am also not sure converting the input to string is a good approach in all cases: what if the user wants to parse a multiple-gigabytes dump? The new code loads it all into memory while the old one was able to parse it in streaming.

If you want you could revert a bit your change and I can give it a try.
Thank you again!

edmondchuc commented:
> I am also not sure converting the input to string is a good approach in all cases: what if the user wants to parse a multiple-gigabytes dump? The new code loads it all into memory while the old one was able to parse it in streaming.

Yes, I agree completely!

> If you want you could revert a bit your change and I can give it a try.


Yeah sure, go for it! I'm keen to learn the correct way. I believe you can push directly to this branch, feel free.

Tpt commented:
The new 0.4.0-alpha.1 version provides this feature: `load("my_string")` will parse `"my_string"` as the file content whereas `load(path="file.ttl")` will parse the `file.ttl` file.
